rfcs/20240312-standardize-tan-customcall-tuplecollective-ops.md

```diff
@@ -0,0 +1,359 @@
+# [RFC] Standardize TanOp, CustomCallOp with typed FFI, and tuple-collectives ops
+
+Status: Draft<br/>
+Initial version: 03/12/2024<br/>
+Last updated: 03/12/2024<br/>
+Discussion thread: [GitHub](add PR Link)
+
+## Motivation
+
+MHLO `tan` op, `custom_call` op with typed FFI and tuple-collectives ops
+(`all_gather`, `all_reduce`, `all_to_all`) supports features which are
+successfully being used by JAX and PT/XLA. Standardizing StableHLO ops
+will ensure HLO-StableHLO feature parity. There are hacks in place to leverage
+these features (unregistered attributes, serialize strings) and standardizing
+the ops is a hack-free solution. Also, there are existing user requests in the
+StableHLO repo for these features.
+
+### TanOp
+
+Frameworks and Compilers both want `tan` op.
+Jax has [`jnp.tan`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.tan.html),
+PyTorch has [`torch.tan`](https://pytorch.org/docs/stable/generated/torch.tan.html).
+On Compilers side, XLA has [`mhlo.tan`](https://github.com/tensorflow/mlir-hlo/blob/master/mhlo/IR/hlo_ops.td#L633).
+StableHLO doesn't support `tan` op. Open ticket for this request
+[#1](https://github.com/openxla/stablehlo/issues/1358)
+
+### CustomCallOp with typed FFI
+
+StableHLO `custom_call` op to support `API_VERSION_TYPED_FFI` as an enum value [`StableHLO_CustomCallApiVersionAttr`](https://github.com/openxla/stablehlo/blob/04365f85cfbffe3d95ba2fb79ff34cd929d4a9a6/stablehlo/dialect/StablehloEnums.td#L88).
```

Comment:
Let's leave the enum value out of this description, since it's mostly implementation detail. I think it's safe to just refer to this as "`custom_call` with a `backend_config` dictionary."